accepted/cranelift-assembler.md

```diff
@@ -0,0 +1,368 @@
+```
+
+When we generate Rust code from the DSL, the assembler creates the following:
+
+```rust
+/// `andl: MI_SXBL(rm32[rw], imm8[sxl]) => 0x83 /4 ib`
+#[derive(arbitrary::Arbitrary, Clone, Debug)]
+pub struct andl_mi_sxbl {
+    pub rm32: GprMem,
+    pub imm8: Imm8,
+}
+impl andl_mi_sxbl {
+    pub fn encode(&self, buf: &mut impl CodeSink, off: &impl KnownOffsetTable) { ... }
+    pub fn matches(&self, features: TODO) { ... }
+    pub fn regalloc(&mut self, visitor: &mut impl RegallocVisitor) {
+        self.rm32.read_write(visitor); // cranelift/assembler/meta/src/generate/inst.rs:130
+    }
+}
+impl std::fmt::Display for andl_mi_sxbl {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        let rm32 = self.rm32.to_string(Size::Doubleword); // cranelift/assembler/meta/src/generate/inst.rs:149
+        let imm8 = self.imm8.to_string(Extension::SignExtendLong); // cranelift/assembler/meta/src/generate/inst.rs:149
+        write!(f, "andl {imm8}, {rm32}") // cranelift/assembler/meta/src/generate/inst.rs:154
+    }
+}
+```
+
+Note how, unlike in the current `Inst` implementation, the types of `andl_mi_sxbl` prevent
+accidental misuse: we can only create this instruction in a correct state. Beyond construction, the
+actual implementation of `encode` is already tested: to a large degree, the prototype reuses the
+pre-existing assembler logic in `cranelift-codegen`. To complete the bridge from `cranelift-codegen`
+to this new assembler, various traits like `CodeSink`, `KnownOffsetTable`, and `RegallocVisitor`
+adapt between what currently exists and the more restrictive assembler. And, for troubleshooting,
+generated code is commented with the location it was generated from.
+
+The generator also creates one large `enum Inst` to comprehend all defined instructions. Functions
+for encoding, matching, pretty-printing, etc., all proxy on to their previously generated versions
+(only pretty-printing shown here):
+
+```rust
+pub enum Inst {
+    andl_mi_sxbl(andl_mi_sxbl),
+    ...
+}
+impl std::fmt::Display for Inst {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::andl_mi_sxbl(i) => write!(f, "{i}"), // cranelift/assembler/meta/src/generate/inst.rs:43
+            ...
+        }
+    }
+}
+```
+
+One final piece to this is the generation of ISLE helpers to ease with integration (see
+[Integration] below). This is still a work in progress in the [`abrown#assembler-collaborate`]
+prototype, but one can imagine a `assembler.isle` file containing definitions like:
+
+```
+(decl x64_andl_mi_sxbl (AssemblerGprMem AssemblerImm8) AssemblerGprMem)
+(rule (x64_andl_mi_sxbl rm32 imm8)
+      (let ((_ Unit (emit (MInst.External (assemble_x64_andl_mi_sxbl rm32 imm8)))))
+        rm32))
+(decl assemble_x64_andl_mi_sxbl (AssemblerGprMem AssemblerImm8) AssemblerInst)
+(extern constructor assemble_x64_andl_mi_sxbl assemble_x64_andl_mi_sxbl)
```

Comment:
One thing I'd be curiuos to see, especially in connection with the verbosity point called out below, is how this would integrate into the lowering of `band` itself at a high level. For example the lowerings we have today feel "right" to me in that they use `(x64_and ...)` as the result. I think it'd be reasonable to rename `x64_and` to `x64_andl_mi_xsbl ...` or such to avoid shoehorning so many instructions into one shape (which I think is a good idea), but with this ISLE snippet I don't know what the output of `AssemblerGprMem` is and how that'd get integrated into the `band` lowering.

Reply:
Yes, this is something discussed below in the "verbosity" bullet. Let me provide more detail by just focusing in on the 8-bit version of CLIF `and`; in `cranelift-codegen`, we have three possibilities for each of the operands, `x` and `y`, to this CLIF instruction. I wrote out all the possible options and thought through how we might lower them:

```
| size | ~x~ | ~y~ | asm             | notes                      | rule |
|------+-----+-----+-----------------+----------------------------+------|
| I8   | Gpr | Gpr | andb_rm/andb_mr |                            |    0 |
| I8   | Gpr | Mem | andb_rm         |                            |    1 |
| I8   | Gpr | Imm | andb_mi/andb_i? |                            |    2 |
| I8   | Mem | Gpr | andb_rm flipped |                            |    3 |
| I8   | Mem | Mem |                 | No, cannot encode this     |      |
| I8   | Mem | Imm | andb_mi         | No, x's memory is modified |      |
| I8   | Imm | Gpr | andb_mi flipped |                            |    4 |
| I8   | Imm | Mem | andb_mi flipped | No, x's memory is modified |      |
| I8   | Imm | Imm |                 | No, cannot encode this     |      |
```

Some options we just can't encode in x64. Other's we can't yet (we need to be pattern-match a read+write to a single memory location). But for the five options we _can_ encode, let's write out some ISLE rules:

```
(rule 0 (lower (has_type $I8 (band (ensure_in_vreg x) (ensure_in_vreg y))))
  (x64_andb_rm x y))
(rule 1 (lower (has_type $I8 (band (ensure_in_vreg x) (sinkable_load y))))
  (x64_andb_rm x y))
(rule 2 (lower (has_type $I8 (band (ensure_in_vreg x) (simm32_from_value y))))
  (x64_andb_mi x y))
(rule 3 (lower (has_type $I8 (band (sinkable_load x) (ensure_in_vreg y))))
  (x64_andb_rm y x))
(rule 4 (lower (has_type $I8 (band (simm32_from_value x) (ensure_in_vreg y))))
  (x64_andb_mi y x))
```

Let's pretend that `ensure_in_vreg` means "this is a `Gpr`," `sinkable_load` means "this is a `Mem`," and ~simm32_from_value~ means "this is an `Imm`" (to reuse some existing ISLE helpers). Turns out there are a few other patterns we can probably encode as well that additionally pattern-match on sign extension.

Two observations from this:
- if we lower at this level, it is _very_ verbose (multiply those rules times more sizes!); this is what I'm requesting discussion on in the "verbose" section below
- this is also very clear: one can see that we just can't lower certain patterns as well as what is the precise instruction emitted for each pattern that we can

-----

Hope all of that connects the dots for how one might verbosely lower to the assembler instructions. But, of course, if we want to retain the existing meta-rules like `x64_and` then we can probably generate those somehow as well. I'm fine with either approach.